File: app/paths.py
from datetime import datetime, timezone
from pathlib import Path

from app.config import settings


class PathNotAllowedError(Exception):
    pass


def _resolved_real(p: Path) -> Path:
    try:
        return p.resolve(strict=True)
    except OSError as e:
        raise PathNotAllowedError("Chemin introuvable") from e


def assert_path_under_roots(requested: str) -> Path:
    expanded = requested
    if requested.startswith("~"):
        expanded = str(Path.home() / requested.removeprefix("~/").removeprefix("~"))

    absolute = Path(expanded).resolve()
    try:
        real = _resolved_real(absolute)
    except PathNotAllowedError:
        raise

    for root in settings.workspace_roots:
        try:
            root_real = _resolved_real(Path(root))
        except PathNotAllowedError:
            continue
        if real == root_real or root_real in real.parents:
            return real

    raise PathNotAllowedError("Ce dossier n'est pas dans WORKSPACE_ROOTS")


def _read_git_branch(repo_path: Path) -> str | None:
    head_file = repo_path / ".git" / "HEAD"
    try:
        trimmed = head_file.read_text(encoding="utf-8").strip()
    except OSError:
        return None
    prefix = "ref: refs/heads/"
    if trimmed.startswith(prefix):
        return trimmed[len(prefix) :]
    return trimmed[:7]


def list_subfolders(root_path: str) -> list[dict]:
    safe_root = assert_path_under_roots(root_path)
    try:
        entries = sorted(safe_root.iterdir(), key=lambda p: p.name.lower())
    except OSError:
        return []

    results: list[dict] = []
    for entry in entries:
        if entry.name.startswith("."):
            continue
        try:
            if not entry.is_dir():
                continue
        except OSError:
            continue

        is_git = (entry / ".git").exists()
        branch = _read_git_branch(entry) if is_git else None
        results.append(
            {
                "name": entry.name,
                "path": str(entry),
                "isGit": is_git,
                "branch": branch,
            }
        )
    return results


def folder_meta(folder_path: str) -> dict:
    safe = assert_path_under_roots(folder_path)
    is_git = (safe / ".git").exists()
    branch = _read_git_branch(safe) if is_git else None
    mtime: str | None = None
    try:
        mtime = datetime.fromtimestamp(safe.stat().st_mtime, tz=timezone.utc).isoformat()
    except OSError:
        pass
    return {
        "path": str(safe),
        "name": safe.name,
        "isGit": is_git,
        "branch": branch,
        "mtime": mtime,
    }

File: server/paths.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { config } from "./config.js";

export class PathNotAllowedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PathNotAllowedError";
  }
}

async function resolvedReal(p: string): Promise<string> {
  try {
    return await fs.realpath(p);
  } catch {
    throw new PathNotAllowedError("Chemin introuvable");
  }
}

export async function assertPathUnderRoots(requested: string): Promise<string> {
  const expanded = requested.startsWith("~")
    ? path.join(os.homedir(), requested.replace(/^~\/?/, ""))
    : requested;

  const absolute = path.resolve(expanded);
  const real = await resolvedReal(absolute);

  for (const root of config.workspaceRoots) {
    let rootReal: string;
    try {
      rootReal = await resolvedReal(root);
    } catch {
      continue;
    }
    if (real === rootReal || real.startsWith(rootReal + path.sep)) {
      return real;
    }
  }

  throw new PathNotAllowedError("Ce dossier n’est pas dans WORKSPACE_ROOTS");
}

export async function listSubfolders(rootPath: string): Promise<
  Array<{
    name: string;
    path: string;
    isGit: boolean;
    branch: string | null;
  }>
> {
  const safeRoot = await assertPathUnderRoots(rootPath);
  let entries: string[];
  try {
    entries = await fs.readdir(safeRoot);
  } catch {
    return [];
  }

  const results: Array<{
    name: string;
    path: string;
    isGit: boolean;
    branch: string | null;
  }> = [];

  for (const name of entries.sort((a, b) => a.localeCompare(b))) {
    if (name.startsWith(".")) continue;
    const full = path.join(safeRoot, name);
    let stat;
    try {
      stat = await fs.stat(full);
    } catch {
      continue;
    }
    if (!stat.isDirectory()) continue;

    let isGit = false;
    let branch: string | null = null;
    try {
      await fs.access(path.join(full, ".git"));
      isGit = true;
      branch = await readGitBranch(full);
    } catch {
      /* not a git repo */
    }

    results.push({ name, path: full, isGit, branch });
  }

  return results;
}

async function readGitBranch(repoPath: string): Promise<string | null> {
  try {
    const head = await fs.readFile(path.join(repoPath, ".git", "HEAD"), "utf8");
    const trimmed = head.trim();
    if (trimmed.startsWith("ref: refs/heads/")) {
      return trimmed.slice("ref: refs/heads/".length);
    }
    return trimmed.slice(0, 7);
  } catch {
    return null;
  }
}

export async function folderMeta(folderPath: string): Promise<{
  path: string;
  name: string;
  isGit: boolean;
  branch: string | null;
  mtime: string | null;
}> {
  const safe = await assertPathUnderRoots(folderPath);
  const name = path.basename(safe);
  let isGit = false;
  let branch: string | null = null;
  try {
    await fs.access(path.join(safe, ".git"));
    isGit = true;
    branch = await readGitBranch(safe);
  } catch {
    /* */
  }
  let mtime: string | null = null;
  try {
    const st = await fs.stat(safe);
    mtime = st.mtime.toISOString();
  } catch {
    /* */
  }
  return { path: safe, name, isGit, branch, mtime };
}
